### PR TITLE
Refactor RealPath

### PR DESCRIPTION
**Pull request by ramchale**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Refactor RealPath for #177

### CODE REVIEW

File: docs/book/v3/standard-filters.md [hunk 1]
```diff
@@ -1156,11 +1156,11 @@ $filtered = $filter->filter($path);
 ### Non-Existing Paths
 
 Sometimes it is useful to get paths to files that do n0t exist; e.g., when you
```
Comment (gsteel) [inline docs/book/v3/standard-filters.md]:
```suggestion
Sometimes it is useful to get paths to files that do not exist; e.g., when you
```

File: docs/book/v3/standard-filters.md [hunk 2]
```diff
@@ -1156,11 +1156,11 @@ $filtered = $filter->filter($path);
...
-want to get the real path for a path you want to create. You can then either
-provide a `FALSE` `exists` value at initiation, or use `setExists()` to set it.
+want to get the real path for a path you want to create. You can then
+provide a `FALSE` `exists` option at initiation.
```
Comment (gsteel) [inline docs/book/v3/standard-filters.md]:
```suggestion
provide `false` for the `exists` option during construction.
```

Comment (gsteel) [reply docs/book/v3/standard-filters.md]: Also, whilst the current docs don't conform to this rule - can you please re-format _(just the changes)_ to one sentence per line please. This is a general, un-documented docs style preference, that hopefully we'll add to a style guide one day 😬

File: src/RealPath.php [hunk 3]
```diff
@@ -24,74 +23,28 @@
...
     {
         if (! is_string($value)) {
             return $value;
         }
         $path = (string) $value;
 
-        if ($this->options['exists']) {
+        if ($this->pathMustExist) {
             return realpath($path);
         }
```
Comment (gsteel) [inline src/RealPath.php:43]: Can we get rid of `ErrorHandler::start()` and `stop()` here? I'm not convinced that any warnings/errors are emitted: https://3v4l.org/jT1XV

File: src/RealPath.php [hunk 4]
```diff
@@ -24,74 +23,28 @@
...
-     * If the value provided is non-scalar, the value will remain unfiltered
-     *
-     * @psalm-return ($value is string ? string : mixed)
-     */
     public function filter(mixed $value): mixed
     {
         if (! is_string($value)) {
             return $value;
         }
         $path = (string) $value;
```
Comment (gsteel) [inline src/RealPath.php]: This cast is not necessary

File: src/RealPath.php [hunk 5]
```diff
@@ -24,74 +23,28 @@
...
 
-        if ($this->options['exists']) {
+        if ($this->pathMustExist) {
             return realpath($path);
```
Comment (gsteel) [inline src/RealPath.php]: Typical filter behaviour should be to either return the successfully filtered value or the un-filtered value. This is potentially returning `false`, so I think there should be conditional to return `$value` when `realpath() === false`.

```php
// file_exists will return appropriate result for args such as '/some/../other/../foo/bar'
if ($this->pathMustExist && ! file_exists($value)) {
    return $value;
}

if ($this->pathMustExist) {
    return realpath($value);
}
```

This behaviour change will need a sentence in the migration docs and the tests will need changing too.

File: src/RealPath.php [hunk 6]
```diff
@@ -24,74 +23,28 @@
 
 /**
  * @psalm-type Options = array{
- *     exists: bool,
- *     ...
+ *     exists?: bool,
  * }
- * @template TOptions of Options
- * @extends AbstractFilter<TOptions>
+ * @implements FilterInterface<string|array<array-key, string|mixed>>
```
Comment (gsteel) [inline src/RealPath.php]:
```suggestion
 * @implements FilterInterface<string>
```

The successfully filtered value is a string - therefore the return type of `filter()` will be `mixed|string` - we might be able to narrow this to `non-empty-string` further if its not a massive pain in the ass.

File: src/RealPath.php (hunk 3 above)
Comment (gsteel) [reply src/RealPath.php:43]: Also for all the code below, it remains untested because CI does not run windows 🤔

**Review by gsteel — CHANGES_REQUESTED** [review]
Thanks @ramchale - The windows related shenanigans need some more attention… I'll have a think about running CI on windows, or perhaps also discuss on slack general windows support…

File: docs/book/v3/standard-filters.md (hunk 2 above)
Comment (froschdesign) [reply docs/book/v3/standard-filters.md]: > # Writing Style
> 
> ## One Sentence per-line Method
> 
> Documentation should be written using [the one sentence per-line method](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line). 
> There are a number of excellent reasons for writing in this style, including:
> 
> - It prevents reflows (meaning a change early in the paragraph won’t cause the remaining lines in the paragraph to reposition)
> - You can easily swap/move sentences
> - You can easily separate or join paragraphs
> - You can comment out sentences or add commentary to them
> - You can spot sentences which are too long or sentences that vary widely in length
> - You can spot redundant (and thus mundane) patterns in your writing
> 
> Take the following content as an example.
> 
> ```markdown
> The concrete `TableGateway` object uses constructor injection for getting
> dependencies and options into the instance. The table name and an instance of
> an `Adapter` are all that is required to create an instance. Out of the box, this 
> implementation makes no assumptions about table structure or metadata, and 
> when `select()` is executed, a simple `ResultSet` object with the populated 
> `Adapter`'s `Result` (the datasource) will be returned and ready for iteration.
> ```
> 
> While it can be easily read in any editor, as it does not require the reader to scroll, that simplicity comes at the cost of maintainability. 
> For example, if you needed to put the second sentence before the first, or move it to any other location, it would need to be cut out of the surrounding paragraph before it could be moved.
> 
> Compare that with the following sample content.
> 
> ```markdown
> The concrete `TableGateway` object uses constructor injection for getting dependencies and options into the instance. 
> The table name and an instance of an `Adapter` are all that is required to create an instance.
> Out of the box, this implementation makes no assumptions about table structure or metadata, and when `select()` is executed, a simple `ResultSet` object with the populated `Adapter`'s `Result` (the datasource) will be returned and ready for iteration.
> ```
> 
> Yes, the content, by default scrolls out of view. 
> However, most text editors and IDEs support "_soft wrapping_", making this a non-issue. 
> Using this approach, the sentence doesn't need to be extracted from the surrounding paragraph before it can be moved.

Comment (gsteel) [reply docs/book/v3/standard-filters.md]: Sorry Frank! the _documented_ style guide… Where is it?

Comment (ramchale) [reply docs/book/v3/standard-filters.md]: Will do. Thanks both

File: src/RealPath.php (hunk 3 above)
Comment (ramchale) [reply src/RealPath.php:43]: > Can we get rid of `ErrorHandler::start()` and `stop()` here? I'm not convinced that any warnings/errors are emitted: https://3v4l.org/jT1XV

Looks like we can. I played around with the open_baseline restriction just to make sure it wasn't something to do with that, but it also just returns false if outside of the allowed directories.

File: src/RealPath.php (hunk 5 above)
Comment (ramchale) [reply src/RealPath.php]: I think there's a strong possibility of some people relying on this returning false for validation (which I know it isn't intended for 😄). As the behavior for passing in a path that doesn't exist would change to returning the same path back, with no indication as to whether that's because it matches the realpath, is it okay to suggest introducing Laminas\Validator\File\Exists in the docs if people aren't using it?

Comment (gsteel) [reply src/RealPath.php]: Yes! Improving docs to state that filters shouldn't be used for validation is a worthy improvement! As filter does not depend on validator, I don't think its _necessary_ to mention laminas-validator itself, but write whatever makes most sense to the user - Typically filtering happens prior to validation, so returning the un-modified path argument when it is un-filterable is good, because it should be validated by `..\File\Exists` afterwards.

**Comment by gsteel** [comment]
Further to the Windows specific code here - can you please remove it, and any related tests?

We should also document dropping windows support in this filter.

File: src/RealPath.php (hunk 5 above)
Comment (ramchale) [reply src/RealPath.php]: Cool. Thanks

File: docs/book/v3/standard-filters.md [hunk 7]
```diff
@@ -1137,13 +1136,12 @@ The following options are supported for `Laminas\Filter\RealPath`:
...
-to `/./`, `/../` and extra `/` sequences in the input path will be stripped. The
-resulting path will not have any symbolic links, `/./`, or `/../` sequences.
+For any given link or pathname, its absolute path will be returned.
+References to `/./`, `/../` and extra `/` sequences in the input path will be stripped.
+The resulting path will not have any symbolic links, `/./`, or `/../` sequences.
 
-`Laminas\Filter\RealPath` will return `FALSE` on failure, e.g. if the file does not exist. On BSD
-systems `Laminas\Filter\RealPath` doesn't fail if only the last path component doesn't exist, while
-other systems will return `FALSE`.
+`Laminas\Filter\RealPath` will return `FALSE` on failure, e.g. if the file does not exist.
```
Comment (gsteel) [inline docs/book/v3/standard-filters.md]:
```suggestion
`Laminas\Filter\RealPath` will return the un-filtered input on failure, for example, if the input is not a string, the value cannot be interpreted as a file path, or, the path does not exist when existence is a configured requirement.
```

File: docs/book/v3/standard-filters.md [hunk 8]
```diff
@@ -1137,13 +1136,12 @@ The following options are supported for `Laminas\Filter\RealPath`:
...
+On BSD systems `Laminas\Filter\RealPath` doesn't fail if only the last path component doesn't exist, while other systems will return `false`.
```
Comment (gsteel) [inline docs/book/v3/standard-filters.md]: Can we verify this? also, it shouldn't return false.

File: src/RealPath.php [hunk 9]
```diff
@@ -42,48 +29,18 @@ public function filter(mixed $value): mixed
...
-                [, $drive, $path] = $matches;
-            } else {
-                $cwd   = getcwd();
-                $drive = substr($cwd, 0, 2);
-                if (! str_starts_with($path, DIRECTORY_SEPARATOR)) {
-                    $path = substr($cwd, 3) . DIRECTORY_SEPARATOR . $path;
-                }
-            }
-        } elseif (! str_starts_with($path, DIRECTORY_SEPARATOR)) {
-            $path = getcwd() . DIRECTORY_SEPARATOR . $path;
```
Comment (gsteel) [inline src/RealPath.php:118]: Conditionally preprending `getcwd()` is a feature we should probably keep and test:

```php
$filter->filter('foo.txt'); // "/var/www/project/foo.txt"
```

File: src/RealPath.php [hunk 10]
```diff
@@ -42,48 +29,18 @@ public function filter(mixed $value): mixed
...
-        $stack = [];
-        $parts = explode(DIRECTORY_SEPARATOR, $path);
-        foreach ($parts as $dir) {
-            if ($dir !== '' && $dir !== '.') {
-                if ($dir === '..') {
-                    array_pop($stack);
-                } else {
-                    $stack[] = $dir;
-                }
-            }
```
Comment (gsteel) [inline src/RealPath.php]: Dropping this code means that you cannot fabricate the `realpath` for a non-existent file, i.e. `realpath('/not-there') === false`

**Review by gsteel — CHANGES_REQUESTED** [review]
Thanks @ramchale, but, we still need to "make up" the real path for non-existent files when appropriate

**Comment by ramchale** [comment]
> Thanks @ramchale, but, we still need to "make up" the real path for non-existent files when appropriate

🤦 sorry. Not my best work. I've reworked some of the tests to better cover that with the return value changes.
I've added a test that should check whether the statement about BSD is true, but don't have access to a BSD machine to check and the docker support attempts seem to have petered out.

**Review by gsteel — CHANGES_REQUESTED** [review]
Thank you @ramchale - this is looking _much_ better - could we improve the test `testExistingFileReturnsRealPath` with a data provider along the lines of

```php
[
   __DIR__ . '/_files/foo/../bar/../file.1',
   __DIR__ . '/_files/././file.1',
   __DIR__ . '///_files///file.1',
   // etc…
]
```

**Comment by ramchale** [comment]
Thanks @gsteel . I've added those.
`__DIR__ . '/_files/foo/../bar/../file.1'` is invalid, but seemed like the sort of edge-case that might come up, so I've dropped it in it's own test.
Once you're happy with the changes, is there any issue with me squashing this before it's merged to cut down on some of the undoing of changes in the commit history?

**Review by gsteel — APPROVED** [review]
Thanks again @ramchale - Great improvement

Let me know when you're done squashing and I'll merge 👍

**Comment by ramchale** [comment]
@gsteel Thanks. Should be ready now

**Review by gsteel — APPROVED** [review]
Thanks!
